docs/html/man/man3/PrefixSettings.3
.TH "PrefixSettings" 3 "8 Jul 2009" "Version 0.113" "q4wine" \" -*- nroff -*-
.ad l
.nh
.SH NAME
PrefixSettings \- .IP "\(bu" 2

.PP
 

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <src/gui/prefixsettings.h>\fP
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBPrefixSettings\fP (QString prefix_name, QWidget *parent=0, Qt::WFlags f=0)"
.br
.RI "\fIThis is class constructor. \fP"
.in -1c
.SS "Private Slots"

.in +1c
.ti -1c
.RI "void \fBcmdCancel_Click\fP ()"
.br
.RI "\fICancle button click slot \fP"
.ti -1c
.RI "void \fBcmdOk_Click\fP ()"
.br
.RI "\fIOk button click slot \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "bool \fBeventFilter\fP (QObject *obj, QEvent *event)"
.br
.RI "\fIThis is an event filter. \fP"
.ti -1c
.RI "void \fBloadThemeIcons\fP (QString themePath)"
.br
.RI "\fIThis function loads theme images to widgets. \fP"
.ti -1c
.RI "QIcon \fBloadIcon\fP (QString iconName, QString themePath)"
.br
.RI "\fIThis function returns load theme image item by user selected theme. \fP"
.ti -1c
.RI "typedef void * \fBCoreLibPrototype\fP (bool)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "QString \fBprefix_id\fP"
.br
.RI "\fI\fBPrefix\fP id, and name. \fP"
.ti -1c
.RI "QString \fBprefix_name\fP"
.br
.ti -1c
.RI "\fBPrefix\fP * \fBdb_prefix\fP"
.br
.RI "\fIDatabase prefix class defenition. \fP"
.ti -1c
.RI "QLibrary \fBlibq4wine\fP"
.br
.RI "\fIThis is need for libq4wine-core.so import. \fP"
.ti -1c
.RI "CoreLibPrototype * \fBCoreLibClassPointer\fP"
.br
.ti -1c
.RI "\fBcorelib\fP * \fBCoreLib\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
.IP "\(bu" 2

.PP


This class provide prefix settings dialog functions. 
.PP
Definition at line 60 of file prefixsettings.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "PrefixSettings::PrefixSettings (QString prefix_name, QWidget * parent = \fC0\fP, Qt::WFlags f = \fC0\fP)"
.PP
This is class constructor. 
.PP
\fBParameters:\fP
.RS 4
\fIprefix_name\fP Current prefix name. 
.RE
.PP

.PP
Definition at line 32 of file prefixsettings.cpp.
.SH "Member Function Documentation"
.PP 
.SS "bool PrefixSettings::eventFilter (QObject * obj, QEvent * event)\fC [private]\fP"
.PP
This is an event filter. 
.PP
This event filter hendle button click events 
.PP
\fBParameters:\fP
.RS 4
\fIprefix_name\fP Current prefix name. 
.RE
.PP

.PP
Definition at line 163 of file prefixsettings.cpp.
.SS "QIcon PrefixSettings::loadIcon (QString iconName, QString themePath)\fC [private]\fP"
.PP
This function returns load theme image item by user selected theme. 
.PP
\fBParameters:\fP
.RS 4
\fIiconName\fP icon name 
.br
\fIthemePath\fP is a path to user selected theme ;) 
.RE
.PP

.PP
Definition at line 120 of file prefixsettings.cpp.
.SS "void PrefixSettings::loadThemeIcons (QString themePath)\fC [private]\fP"
.PP
This function loads theme images to widgets. 
.PP
\fBParameters:\fP
.RS 4
\fIthemePath\fP is a path to user selected theme ;) 
.RE
.PP

.PP
Definition at line 99 of file prefixsettings.cpp.

.SH "Author"
.PP 
Generated automatically by Doxygen for q4wine from the source code.


ChangeLog
Installation
***********************************************
See INSTALL file for details

ChangeLog
***********************************************

Version 0.113
 Added:
  - Added embeded q4wine-mount (This is a copy of fuseiso + RH path);
       Note: use this if you too lazy to compile fuseiso from SF and apply pathes;
  - fuseiso ( http://sourceforge.net/projects/fuseiso/ ) full support.
  - q4wine now remembers 8 recent mounted images;
  - q4wine now remembers 8 recent runned binaryes via Run dialog;
  - QtSingleApp integration. Now you can run only one instance of q4wine-gui;
  - Online documentation;
  - New website: http://q4wine.brezblock.org.ua/;
  - Now q4wine save last user selected prefix and dir;
  - Added "open directory" menu items via xdg-utils;
  - Added "open directory" menu items via winefile;
  - Added xdg-utils support (note: now it is in the depends list)
  - Added q4wine-cli console utility for wine applications and prefixes management.
  - Added libq4wine implementation;
  - Added translation file for Portuguese (Brazil) by Marcio Moraes;
  - Added *.xpm filter to icon import patten;
  - Added http proxy support for winetriks;
  - Improved icon display widget;
  - Added Drag & Drop support;
  - Added Drag support q4wine icons export;
  - Added Drop support for wine .exe and .com files;
  - Added Drop support for wine .bat files (Now autoadd wineconsole binary args);
  - Added splitter for programs and icons lists;
  - Added command line option for q4wine (See q4wine --help for details);

 Fixed:
  - Linux: Improved wine process list build (thx Sergey Kishchenko (tilarids));
  - Cleanup q4wine tmp directory on exit;
  - Some fixes for q4wine.desktop (thx to Eugene Pivnev);
  - Fixed GUI bug: Text fields size, on some desktop configurations, are too small to edit;
  - Fixed a lot of English spelling errors (thx Sergey Kishchenko (tilarids));
  - First steps for source code documentation via Doxygen....;
  - Total code reorganize;
  - Database engine rewrite;
  - Many fixes for q4wine.desktop file (thx for Kyrill Detinov);
  - Fixed compilation with Qt-4.4.X (thx for Kyrill Detinov);

Version 0.112-r1
 Fixed:
  - Fixed compilation with Qt-4.4.X;
  - Removed QDir::Files from Directory selection dialogs;
  - Fixed compilation warning for registry.h;
  - Fixed warnings ignoring return value of 'int system(const char*)' on 1666 and 1647 lines;

Version 0.112
 core:
  New:
  - Added Sidebar Urls for Run and IconSettings file dialogs;
  - Wine Dlls path autodetections; (If not work on your system -- plz mail to me ;) )
  - You can disable icotools support in q4wine by adding -DWITH_ICOTOOLS=OFF to cmake;
  - Winetools support is disabled by default. You can enable it by adding -DWITH_WINETOOLS=ON to cmake args;
  - All developer stuff is disabled by default. You can enable it by adding -DWITH_DEVELOP_STUFF=ON to cmake args;
  - New database class;
  - kdesu\gksu support;
  Fixed:
  - Fixed import\export prefix functions;
  - Many small fixes;
  - Updated STDERR message catch function;
  - Added FreeBSD port patches by Dmitry Marakasov aka amdmi3;
  - Fixed readonly text field on prefix create;
  - Fixed q4wine.desktop GENERIC names;
  - Fixed IconSettings path detections (ProgramPath\WorkDirPath e.t.c.);
  - Fixed .exe adding to IconName on Prgoram file select;
  - Run, AppSettings, ImageManager, Wizard and IconSettings dialogs now are modal dialogs;
  - Fixed LC_ALL and LS_MESSAGES locale detection;
  - Winetriks executable now puts into $HOME/.config/q4inwe/ directory;
  - No sudo now needed for Winetriks download and install;

Version 0.111
 core:
  New:
  - New q4wine icons with different sizes;
  - Updated q4wine.desktop;
  - Make install now install q4wine.desktop and icons;
  - Added basic shortcuts;
  - q4wine now save and restore size and position of main window;
 core:
  Fixed:
  - Fixed icoutils icon export bugs;
  - Fixed file name case sense on icon exort;
  - Added .exe or .dll file selection masks on icon export;
  - Fixed start selection file path on icon export;
  - FreeBSD: kvm functions now do not need kmem gid;
  - FreeBSD: wine_pthread now resolve exec name;
  - Updated About dialog;
  - Updated Ukrainian and Russian locales;

Version 0.110
 core:
  New:
  - FreeBSD full support (proc and mounts);
  - nrg images mount (Linux only);
  - new core functions class (will be moved to libq4wine shared library);
  - updated Ukrainian and Russian locales;
  - basic wis xml parser class;
  - Added nice level field to Run dialog;
  - Added desktop size to Run dialog;

  Fixed:
  - Winetriks prefix fixes;
  - FreeBSD kmem read exception check;
  - Can't delete image from ImageManager;
  - Table view and resize;
  - Fixed getWineDll at Run dialog;
  - User Locale detection bug;
  - Copy & Paste issues;
  - Removed deprecated konsole args;
  - Fixed Prefix import bug;
  - Fixed /bin/tar compiled path for import and export prefixes;
  - Many fixes;
  - Fixed CMake qm locale gen\install;

Version 0.109-r2
 core:
  New:
  - New CMake (required version 2.6)) build system now used;
  - Added q4wine.desktop file;
  - New core class;

  Experimental:
  - Temporary winetricks support;
  - Added winetricks download support;

  Fixed:
  - Fixed icon options edit error: "Debug: Parameter count mismatch";
  - Fixed regression at icon options edit;
  - Removed q4wine.pro;
  (Note: qmake now deprecated, use cmake)
  - updated INSTALL text file;
  - Fixed empty console arguments process;

Version 0.109-r1
 core:
  New:
  - Russian translation by Pavel Zinin (pashazz)

  Fixed:
  - Fixed icon add error: "Debug: Parameter count mismatch";
  - Fixed spaces in Program path and Working directory errors;
  - Removed src.pro;
  - Added INSTALL text file;

Version 0.109
 gui
  New:
  - New CD image manager;
  - New CD mount menu style;
  - TrayIcon now shows and hide q4wine when double clicking on it;

  Fixed:
  - Removed mount point and cd image settings at Icon Settings;

src/gui/prefixsettings.cpp
/***************************************************************************
 *   Copyright (C) 2008, 2009 by Malakhov Alexey                           *
 *   brezerk@gmail.com                                                     *
 *                                                                         *
 *   This program is free software: you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation, either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
 *                                                                         *
 *   In addition, as a special exception, the copyright holders give       *
 *   permission to link the code of this program with any edition of       *
 *   the Qt library by Trolltech AS, Norway (or with modified versions     *
 *   of Qt that use the same license as Qt), and distribute linked         *
 *   combinations including the two.  You must obey the GNU General        *
 *   Public License in all respects for all of the code used other than    *
 *   Qt.  If you modify this file, you may extend this exception to        *
 *   your version of the file, but you are not obligated to do so.  If     *
 *   you do not wish to do so, delete this exception statement from        *
 *   your version.                                                         *
 ***************************************************************************/

#include "prefixsettings.h"

PrefixSettings::PrefixSettings(QString prefix_name, QWidget * parent, Qt::WFlags f) : QDialog(parent, f)
{
	// Setup base UI
	setupUi(this);

	// Setting class prefix name
	this->prefix_name=prefix_name;

	// Loading libq4wine-core.so
	libq4wine.setFileName("libq4wine-core");

	if (!libq4wine.load()){
		libq4wine.load();
	}

	// Getting corelib calss pointer
	CoreLibClassPointer = (CoreLibPrototype *) libq4wine.resolve("createCoreLib");
	CoreLib = (corelib *)CoreLibClassPointer(true);

	// Creating database classes
	db_prefix = new Prefix();

	this->loadThemeIcons(this->CoreLib->getSetting("app", "theme", false).toString());

	QStringList result = db_prefix->getFieldsByPrefixName(prefix_name);
	if (result.at(0) == "-1")
		return;

	prefix_id=result.at(0);

	txtWineLibs->setText(result.at(2));
	txtWineLoaderBin->setText(result.at(3));
	txtWineServerBin->setText(result.at(4));
	txtWineBin->setText(result.at(5));

	//comboDeviceList
	txtMountPoint->setText(result.at(6));

	if (prefix_name=="Default"){
		txtPrefixName->setEnabled(FALSE);
		txtPrefixPath->setEnabled(FALSE);
		cmdGetPrefixPath->setEnabled(FALSE);
	} else {
		cmdGetPrefixPath->installEventFilter(this);
	}

	txtPrefixName->setText(prefix_name);
	txtPrefixPath->setText(result.at(1));

	comboDeviceList->addItems(CoreLib->getCdromDevices());
	if (!result.at(7).isEmpty()){
		comboDeviceList->setCurrentIndex (comboDeviceList->findText(result.at(7)));
		if (comboDeviceList->currentText().isEmpty())
			comboDeviceList->setCurrentIndex (0);
	} else {
		comboDeviceList->setCurrentIndex (0);
	}

	connect(cmdCancel, SIGNAL(clicked()), this, SLOT(cmdCancel_Click()));
	connect(cmdOk, SIGNAL(clicked()), this, SLOT(cmdOk_Click()));
	connect(cmdHelp, SIGNAL(clicked()), this, SLOT(cmdHelp_Click()));

	cmdGetWineBin->installEventFilter(this);
	cmdGetWineServerBin->installEventFilter(this);
	cmdGetWineLoaderBin->installEventFilter(this);
	cmdGetWineLibs->installEventFilter(this);

	cmdGetMountPoint->installEventFilter(this);

	return;
}



void PrefixSettings::loadThemeIcons(QString themePath){
	QPixmap pixmap;

	if (!pixmap.load(tr("%1/data/exec.png").arg(themePath))){
		pixmap.load(":data/exec.png");
	}

	lblLogo->setPixmap(pixmap);

	cmdGetWineBin->setIcon(loadIcon("data/folder.png", themePath));
	cmdGetWineServerBin->setIcon(loadIcon("data/folder.png", themePath));
	cmdGetWineLoaderBin->setIcon(loadIcon("data/folder.png", themePath));
	cmdGetWineLibs->setIcon(loadIcon("data/folder.png", themePath));

	cmdGetMountPoint->setIcon(loadIcon("data/folder.png", themePath));
	cmdGetPrefixPath->setIcon(loadIcon("data/folder.png", themePath));

	return;
}


QIcon PrefixSettings::loadIcon(QString iconName, QString themePath){
	// Function tryes to load icon image from theme dir
	// If it fails -> load default from resource file

	QIcon icon;

	if ((!themePath.isEmpty()) and (themePath!="Default")){
		icon.addFile(tr("%1/%2").arg(themePath).arg(iconName));
		if (icon.isNull()){
			icon.addFile(tr(":/%1").arg(iconName));
		}
	} else {
		icon.addFile(tr(":/%1").arg(iconName));
	}

	return icon;
}

void PrefixSettings::cmdCancel_Click(){
	reject();
	return;
}

void PrefixSettings::cmdOk_Click(){
	if (txtPrefixName->text().isEmpty()){
		 QMessageBox::warning(this, tr("Error"), tr("Please, enter prefix name"));
		  return;
	}

	if (prefix_name!=txtPrefixName->text()){
		if (db_prefix->isExistsByName(txtPrefixName->text())){
			QMessageBox::warning(this, tr("Error"), tr("Sorry, but prefix named %1 already exists.").arg(txtPrefixName->text()));
			return;
		}
	}

	if (!db_prefix->updatePrefix(txtPrefixName->text(), txtPrefixPath->text(), txtWineBin->text(), txtWineServerBin->text(), txtWineLoaderBin->text(), txtWineLibs->text(), txtMountPoint->text(), comboDeviceList->currentText(), this->prefix_name))
		reject();

	accept();
	return;
}

bool PrefixSettings::eventFilter(QObject *obj, QEvent *event){
	/*
	 * Select folder dialog
	 */

	if (event->type() == QEvent::MouseButtonPress) {

		QString file;

		if (obj->objectName().right(3)=="Bin"){
			file = QFileDialog::getOpenFileName(this, tr("Open File"), QDir::homePath(),   "All files (*.*)");
		} else {
			file = QFileDialog::getExistingDirectory(this, tr("Open Directory"), QDir::homePath(),   QFileDialog::DontResolveSymlinks);
		}

		if (!file.isEmpty()){
			QString a;

			a.append("txt");
			a.append(obj->objectName().right(obj->objectName().length()-6));

			QLineEdit *lineEdit = findChild<QLineEdit *>(a);

			if (lineEdit){
				lineEdit->setText(file);
			} else {
				qDebug("Error");
			}
		}
	}

	return FALSE;
}

void PrefixSettings::cmdHelp_Click(){
	QString rawurl;
	switch (twbGeneral->currentIndex()){
	case 0:
		rawurl = "10-prefix-settings.html#general";
	break;
	case 1:
		rawurl = "10-prefix-settings.html#winepath";
	break;
	}

	CoreLib->openHelpUrl(rawurl);
}

